lib/src/test/java/io/ably/lib/test/realtime/RealtimeChannelTest.java

```diff
@@ -1019,6 +1021,100 @@ public void detach_success_callback_detaching() throws AblyException {
+     * verify attach {@code CompletionListener#onSuccess()} gets called.
+     */
+    // Spec: RTL4h
+    // https://github.com/ably/ably-java/issues/885
+    @Test
+    public void attach_when_channel_in_detaching_state() throws AblyException {
+        AblyRealtime ably = null;
+        try {
+            ClientOptions opts = createOptions(testVars.keys[0].keyStr);
+            opts.logLevel = Log.VERBOSE;
+            ably = new AblyRealtime(opts);
+
+            /* wait until connected */
+            (new ConnectionWaiter(ably.connection)).waitFor(ConnectionState.connected);
+            assertEquals("Verify connected state reached", ably.connection.state, ConnectionState.connected);
+
+            /* create a channel and attach */
+            final String channelName = "attach_channel";
+            final Channel channel = ably.channels.get(channelName);
+            channel.attach();
+            new ChannelWaiter(channel).waitFor(ChannelState.attached);
+            assertEquals("Verify attached state reached", channel.state, ChannelState.attached);
+
+            /* detach */
+            final Helpers.CompletionWaiter detachCompletionWaiter = new Helpers.CompletionWaiter();
+            channel.detach(detachCompletionWaiter);
+            assertEquals("Verify detaching state reached", channel.state, ChannelState.detaching);
+            final Helpers.CompletionWaiter attachCompletionWaiter = new Helpers.CompletionWaiter();
+            //attempt to attach while detaching
+            channel.attach(attachCompletionWaiter);
+
+            /* Verify onSuccess callback gets called */
+            detachCompletionWaiter.waitFor();
+            assertThat(detachCompletionWaiter.success, is(true));
+            assertThat(channel.state, is(ChannelState.detached));
+            //verify reattach - after detach
+            attachCompletionWaiter.waitFor();
+            assertThat(attachCompletionWaiter.success,is(true));
+            assertThat(channel.state, is(ChannelState.attached));
+        } finally {
+            if(ably != null)
+                ably.close();
+        }
+    }
+
+    /**
+     * When client detaches from a channel in attaching state, verify that detach call will be done after attach
+     * response is received
+     * verify attach {@code CompletionListener#onSuccess()} gets called.
+     */
+    // Spec: RTL5i
+    // https://github.com/ably/ably-java/issues/885
+    @Test
+    public void detach_when_channel_in_attaching_state() throws AblyException {
+        AblyRealtime ably = null;
+        try {
+            ClientOptions opts = createOptions(testVars.keys[0].keyStr);
+            opts.logLevel = Log.VERBOSE;
+            ably = new AblyRealtime(opts);
+
+            /* wait until connected */
+            (new ConnectionWaiter(ably.connection)).waitFor(ConnectionState.connected);
+            assertEquals("Verify connected state reached", ably.connection.state, ConnectionState.connected);
```

Comment:
Minor observation, and I recognise that lots of other sites in this codebase do this wrong, but the JUnit API wants `expected` first, followed by `actual`, as this ordering is used in some of the log messages it outputs. Therefore, for this line:

```suggestion
            assertEquals("Verify connected state reached", ConnectionState.connected, ably.connection.state);
```

This also applies to a few other asserts in this test method.

Reply:
I had also realized this elsewhere and I seem to apply the same mistakes. I have change orders d0d8665517582da836d7e41ac2ac8f95851eb0f4 and created an issue to fix others: https://github.com/ably/ably-java/issues/892